### DISCUSSION TITLE
Bootloader, application & Pelion

### DISCUSSION THREAD

**Discussion by ladislas** [ARMmbed/mbed-tools#187] [Ideas]
Hi there --

I would like to start the discussion about how (and when) the new tools will handle working with a bootloader, an application and ultimately be compatible with Pelion.

The new CMake backend is a pleasure to work with in CI, so it makes having those feature even more important, so that we can add CD to our CI and automatically push new updates.

In the case where support for this is not short term, I would love to know how we cans still use the old tools, or even part of the tools, just the python scripts, to actually make that work.

I have no problem compiling the bootloader on one side, the app on the other and doing the merge "manually" with custom scripts. I'd just like to know where to start.

The Pelion and bootloader examples are great but there are dozens of options to set in `mbed_app.json`, I don't know how they are used, if they are used during compilation? linking? just by the tools to do the merge?

Thanks a lot for the help! 

Best,  
-- Ladislas

**Comment by Patater** [comment]
We haven't given this much thought, yet. How would you like it to work? What's the ideal workflow look like for you? Are you trying to follow some existing documentation for Pelion or Mbed Bootloader and wanting to replicate that experience identically? (If so, please send a link to the documentation you have in mind.)

**Comment by ladislas** [comment]
Thanks for your message @Patater ! :) 

Regarding documentation, we've followed instructions from:

- https://github.com/PelionIoT/mbed-bootloader
- https://github.com/ARMmbed/mbed-os-example-pelion

It works great but as you see it uses only the old tools.

Regarding workflow, what we would like:

- build our firmware to use locally in our products with STLink flashing -- that's what we do today, we don't want to change that
- build the bootloader with the old tools (from @teetak01's comment https://github.com/PelionIoT/mbed-bootloader/issues/95#issuecomment-778005109, it is not planned to use the new tools yet) -- it's okay for us to do so as the bootloader is really only compiled once and won't change when the product ship
- build our firmware to merge with the bootloader to flash our product during production
- build just the application to send updates with Pelion -- this is done automatically with the old tools

The two last parts are the one that we need to get working quite urgently.

The relevant part in the old tools seems to be here: https://github.com/ARMmbed/mbed-os/blob/master/tools/regions.py

**Reply by teetak01** [reply]
Hi @ladislas we do have something that might work as an interim solution at least. We already have a python script that can do the merging.

In the early days before Mbed OS supported the automatic bootloader merging we used this kind of flow for Mbed OS.

If you look to our tutorial for Client 1.5.0 release you can see an example: https://developer.pelion.com/docs/device-management/v1.5/connecting/qs-bootloader.html

and take the old version the example application https://github.com/PelionIoT/mbed-cloud-client-example/tree/1.5.0/tools

Will still use this kind of flow for non-Mbed OS embedded systems.

For. example: https://github.com/PelionIoT/mbed-cloud-client-example/blob/4.7.1/pal-platform/SDK/LPCXpresso54628/tools/combine_bootloader_with_app.py

Similarly the update-image just is the same base image without bootloader and the header information the script (or Mbed tools) is writing.

**Reply by ladislas** [reply]
Thanks a lot @teetak01! It looks really promising, we'll deep dive in all that with the team next week :)
